### PR TITLE
Add SELU activation function

### LINKED ISSUE #1768: Add SELU (Scaled ELU)
Here's TF example:
https://github.com/bioinf-jku/SNNs/blob/master/selu.py#L20-L25

Ideally it would be done in C in TH backend.
 
However, it could also be done in python if a pytorch equivalent of np.where was available:
https://discuss.pytorch.org/t/how-can-i-do-the-operation-the-same-as-np-where/1329/3

### PR DESCRIPTION
**Pull request by fmassa**
Fixes https://github.com/pytorch/pytorch/issues/1768

Also fixes a problem with RReLU in inplace mode.

### CODE REVIEW

File: torch/nn/_functions/thnn/activation.py [hunk 1]
```diff
@@ -95,7 +98,47 @@ def backward(self, grad_output):
...
+        )
+        return grad_input
+
+
+class SELU(InplaceFunction):
+
+    def __init__(self, inplace=False):
+        super(SELU, self).__init__(inplace)
+        self.alpha = 1.6732632423543772848170429916717
+        self.scale = 1.0507009873554804934193349852946
```
Comment (apaszke) [inline torch/nn/_functions/thnn/activation.py]: These can be class attrs. Also, can you write it as a new-style Function?

File: torch/nn/modules/activation.py [hunk 2]
```diff
@@ -245,6 +245,40 @@ def __repr__(self):
             + inplace_str + ')'
 
 
+class SELU(Module):
+    """Applies element-wise, :math:`f(x) = scale * (max(0,x) + min(0, alpha * (exp(x) - 1)))`,
+    with alpha=1.6732632423543772848170429916717 and scale=1.0507009873554804934193349852946.
+    More details can be found in the paper `Self-Normalizing Neural Networks`_ .
+
+    Args:
+        inplace: can optionally do the operation in-place
```
Comment (apaszke) [inline torch/nn/modules/activation.py]: `inplace (bool, optional): ...`

**Review by apaszke — COMMENTED** [review]
All contbuilds are failing

**Comment by fmassa** [comment]
Yeah, I just saw that the tests are failing because there is no `legacy` equivalent for SELU. Should we add it or just skip on SELU?

**Comment by apaszke** [comment]
No, you only have to move the test spec from `common_nn.py` to `test_nn.py`

File: torch/nn/_functions/thnn/activation.py (hunk 1 above)
Comment (fmassa) [inline torch/nn/_functions/thnn/activation.py]: I'm not sure we want to allow the user to specify them. The reason why they chose these parameters is to enforce that the activations have on average zero mean and unit std.
I'll modify it to be new-style

Comment (apaszke) [inline torch/nn/_functions/thnn/activation.py]: I didn't mean that they can be user-specified, but that they don't need to be set in `__init__`. You can put the `alpha = ...` in the class body, and it makes `__init__` completely unnecessary in this case

Comment (fmassa) [inline torch/nn/_functions/thnn/activation.py]: Ah ok, I misunderstood your statement. I'm fixing it

**Comment by fmassa** [comment]
Added new-style Function to `SELU`. Double backprop was verified with `gradgradcheck`.

**Review by apaszke — APPROVED** [review]

File: torch/nn/_functions/thnn/activation.py [hunk 3]
```diff
@@ -95,11 +98,57 @@ def backward(self, grad_output):
...
             self.train,
-            self.inplace
+            False
         )
         return grad_input
 
 
+class SELU(InplaceFunction):
+    alpha = 1.6732632423543772848170429916717
+    scale = 1.0507009873554804934193349852946
```
Comment (andreh7) [inline torch/nn/_functions/thnn/activation.py]: would it make sense to add a comment that these two values are fixed point solutions to equations (4) and (5) in the paper (arxiv:1706.02515) for input/output mean zero and unit variance and a normalized weight vector and that their analytical expression is given in equation (14) ?
